drivers/sensor/ms5611/ms5611.c

```diff
@@ -0,0 +1,551 @@
+		LOG_ERR("Failed to request pressure read");
+		return ret;
+	}
+
+	ret = i2c_read_dt(&cfg->i2c_bus, press_data, sizeof(press_data));
+	if (ret < 0) {
+		LOG_ERR("Failed to read pressure measurement");
+		return ret;
+	}
+
+	raw_press = sys_get_be24(press_data);
+
+	if (raw_press == 0) {
+		LOG_DBG("Invalid pressure data obtained");
+		return -EIO;
+	}
+
+	/* Calculate compensated temperature value */
+	dT = raw_temp - (data->prom[5] << 8);
+	temp = 2000 + (((int64_t)dT * data->prom[6]) >> 23);
+
+	/* Second order temperature compensation */
+	if (temp < 2000) {
+		t2 = ((dT ^ 2) >> 31);
+		off2 = 5 * ((temp - 2000) ^ 2) / 2;
+		sens2 = 5 * (temp - 2000) ^ 2 / 4;
+
+		if (temp < -1500) {
+			off2 = off2 + (7 * (temp + 1500) ^ 2);
+			sens2 = sens2 + (11 * ((temp + 1500) ^ 2)) / 2;
+		}
+	} else {
+		t2 = 0;
+		off2 = 0;
+		sens2 = 0;
+	}
+
+	/* Calculate values with respect to offsets */
+	temp = temp - t2;
+
+	off = (uint32_t)(data->prom[2] << 16) +
+	      (((int64_t)data->prom[4] * dT) >> 7);
+	off = off - off2;
+
+	sens = (data->prom[1] << 15) + (((int64_t)data->prom[3] * dT) >> 8);
+	sens = sens - sens2;
+
+	p = (((uint64_t)(raw_press * sens) >> 21) - off) >> 15;
+
+	data->meas.temp = temp;
+	/* 10mbar = 1kPa */
+	data->meas.press = p / 10;
+
+	return 0;
+}
+
+static int ms5611_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	ret = ms5611_fetch_temp_and_press(dev);
+
+	if (ret < 0) {
+		LOG_ERR("Failed to fetch temperature and pressure");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ms5611_channel_get(const struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct ms5611_data *data = dev->data;
+	struct ms5611_meas_data *meas = &data->meas;
+
+	switch (chan) {
+	case SENSOR_CHAN_PRESS:
+		val->val1 = meas->press / 100;
+		val->val2 = meas->press % 100;
+		break;
+
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		val->val1 = meas->temp / 100;
+		val->val2 = meas->temp % 100;
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int ms5611_fetch_prom(const struct device *dev)
+{
+	const struct ms5611_config *cfg = dev->config;
+	struct ms5611_data *data = dev->data;
+	int ret;
+	uint8_t i;
+	uint8_t i2c_cmd;
+	uint8_t data_bytes[2];
+
+	for (i = 0; i < MS5611_PROM_SIZE; i++) {
+
+		i2c_cmd = MS5611_PROM_READ_BASE;
+
+		/* last bit is not significant */
+		i2c_cmd |= i << 1;
+
+		ret = i2c_write_dt(&cfg->i2c_bus, &i2c_cmd, sizeof(i2c_cmd));
+		if (ret < 0) {
+			LOG_ERR("Failed to request PROM address %d of ms5611",
+				i);
+			return ret;
+		}
+
+		ret = i2c_read_dt(&cfg->i2c_bus, data_bytes,
+				  sizeof(data_bytes));
+
+		if (ret < 0) {
+			LOG_ERR("Failed to get PROM address %d of ms5611", i);
+			return ret;
+		}
+
+		/* convert data to little-endian */
+		data->prom[i] = sys_get_be16(data_bytes);
+	}
+
+	return 0;
+}
+
+static uint8_t ms5611_calculate_crc4(const struct device *dev)
```

Comment:
Can you use a builtin CRC implementation? `include/zephyr/sys/crc.h`

Reply:
Actually, I wanted to do so... Please correct me if I am wrong, but I see in `include/zephyr/sys/crc.h` only functions for 32, 16, 8 and 7 bit CRC calculation. In this sensor datasheet it is stated that we need 4 bit CRC. I didn't found any information how to do that using built in funtions. Is there some other way?

Reply:
if you don't see one you could add to it, unless it's specific to the ms5611 and nothing else

Reply:
According to this application note:
https://www.te.com/commerce/DocumentDelivery/DDEController?Action=showdoc&DocId=Specification+Or+Standard%7FAN520_C-code_example_for_MS56xx%7FA%7Fpdf%7FEnglish%7FENG_SS_AN520_C-code_example_for_MS56xx_A.pdf%7FCAT-BLPS0003
I am afraid that it is custom solution.

Reply:
Please add some comments to the code or commit message explaining why you've implemented a custom CRC

Reply:
Added in the commit message :+1: